File: 2105. Watering Plants II.java
// Medium
// Two Pointers
// Time:O(n),Space:O(1)
// https://leetcode.cn/problems/watering-plants-ii/

class Solution {
  public int minimumRefill(int[] plants, int capacityA, int capacityB) {
      int left = 0, right = plants.length - 1;
      int waterA = capacityA, waterB = capacityB;
      int refills = 0;
      while (left <= right) {
          if (left == right) {
              // if alice and bob meet, the one with more water waters it
              if (Math.max(waterA, waterB) < plants[left]) {
                  refills++;
              }
          } else {
              // alice waters left, bob waters right
              if (waterA < plants[left]) {
                  refills++;
                  waterA = capacityA;
              }
              waterA -= plants[left];
              if (waterB < plants[right]) {
                  refills++;
                  waterB = capacityB;
              }
              waterB -= plants[right];
          }
          left++;
          right--;
      }
      return refills;
  }
}

File: 2001-3000/2105. Watering Plants II.py
# Time:O(n), Space:O(1)

from typing import List

class Solution:
    def minimumRefill(self, plants: List[int], capacityA: int, capacityB: int) -> int:
        n = len(plants)
        left, right = 0, n - 1
        waterA, waterB = capacityA, capacityB
        ans = 0

        while left < right:
            # Alice do left side
            if waterA < plants[left]:
                ans += 1
                waterA = capacityA
            waterA -= plants[left]
            left += 1

            # Bob do right side
            if waterB < plants[right]:
                ans += 1
                waterB = capacityB
            waterB -= plants[right]
            right -= 1
        
        if left == right:
            if waterA >= waterB:
                if waterA < plants[left]:
                    ans += 1
            else:
                if waterB < plants[left]:
                    ans += 1
        
        return ans
